sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,1013 @@
+#[inline(always)]
+pub unsafe fn get_hash_from_slice_unchecked(
+    data: &[u8],
+    target_slot: Slot,
+    num_entries: usize,
+) -> Option<&[u8; HASH_BYTES]> {
+    position_from_slice_binary_search_unchecked(data, target_slot, num_entries).map(|index| {
+        let entry_offset = NUM_ENTRIES_SIZE + index * ENTRY_SIZE;
+        let hash_offset = entry_offset + SLOT_SIZE;
+        let hash_bytes = data.get_unchecked(hash_offset..(hash_offset + HASH_BYTES));
+        &*(hash_bytes.as_ptr() as *const [u8; HASH_BYTES])
+    })
+}
+
+/// Gets a reference to the `SlotHashEntry` at a specific index from a raw byte slice **without validation**.
+///
+/// # Safety
+/// Caller must guarantee `data` contains a valid `SlotHashes` structure and that `index` is less than the entry count derived from the data's prefix.
+#[inline(always)]
+pub unsafe fn get_entry_from_slice_unchecked(data: &[u8], index: usize) -> &SlotHashEntry {
+    let entry_offset = NUM_ENTRIES_SIZE + index * ENTRY_SIZE;
+    let entry_bytes = data.get_unchecked(entry_offset..(entry_offset + ENTRY_SIZE));
+    &*(entry_bytes.as_ptr() as *const SlotHashEntry)
+}
+
+/// Iterator over the entries in `SlotHashes`.
+///
+/// Yields references `&'s SlotHashEntry` tied to the lifetime `'s` of the borrow
+/// of the `SlotHashes` instance.
+pub struct SlotHashesIterator<'s, T>
+where
+    T: Deref<Target = [u8]>,
+{
+    slot_hashes: &'s SlotHashes<T>,
+    current_index: usize,
+}
+
+// Implement Iterator trait for the custom iterator struct
+// TODO: trait extension for unchecked Iterator::next()
+impl<'s, T> Iterator for SlotHashesIterator<'s, T>
+where
+    T: Deref<Target = [u8]>,
+{
+    type Item = &'s SlotHashEntry;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Use safe get_entry method from SlotHashes
+        let entry = self.slot_hashes.get_entry(self.current_index);
+        if entry.is_some() {
+            self.current_index += 1;
+        }
+        entry
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = self.slot_hashes.len().saturating_sub(self.current_index);
+        (remaining, Some(remaining))
+    }
+}
+
+// Implement ExactSizeIterator as we know the exact length
+impl<T> ExactSizeIterator for SlotHashesIterator<'_, T> where T: Deref<Target = [u8]> {}
+
+// Implement IntoIterator for references to SlotHashes
+// This allows using `for entry in &slot_hashes { ... }`
+impl<'s, T> IntoIterator for &'s SlotHashes<T>
+where
+    T: Deref<Target = [u8]>,
+{
+    type Item = &'s SlotHashEntry;
+    type IntoIter = SlotHashesIterator<'s, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        SlotHashesIterator {
+            slot_hashes: self,
+            current_index: 0,
+        }
+    }
+}
+
+#[inline(always)]
+unsafe fn core_binary_search<FFirstSlot, FSlotAtIndex>(
+    target_slot: Slot,
+    num_entries: usize,
+    _get_first_slot: FFirstSlot,
```

Comment:
Why ask for a function that returns the first slot, if it is not needed?

Reply:
An artifact from an old version of search that used average slot decrements to make a very close guess at a first slot to look at. Sadly, it benched on average the same CUs as binary search, since what was gained with reduced search time was lost to the heuristic calculation.

This is removed in a cleanup commit coming shortly.